mongo/description/server_selector.go

```diff
@@ -30,10 +31,52 @@ func (ssf ServerSelectorFunc) SelectServer(t Topology, s []Server) ([]Server, er
 	return ssf(t, s)
 }
 
+// serverSelectorInfo contains metadata concerning the server selector for the
+// purpose of publication.
+type serverSelectorInfo struct {
+	Type           string
+	CustomSelector bool
+	Data           string               `json:",omitempty"`
+	Selectors      []serverSelectorInfo `json:",omitempty"`
+}
+
+// String returns the JSON string representation of the serverSelectorInfo.
+func (sss serverSelectorInfo) String() string {
+	bytes, _ := json.Marshal(sss)
```

Comment:
I'm concerned that the nested JSON created here will be difficult to use when output by a structured logger. Is there another format we can use to combine multiple points of info that works better when nested in JSON? Or maybe only produce JSON when the output is unstructured?

Reply:
Could you elaborate on what you mean by this: "difficult to use when output by a structured logger" ? The output is the type of selector, associated data, and the child selectors (which is only relevant in the composite case). Here is an example:

```json
{
  "Type": "compositeSelector",
  "Selectors": [
    {
      "Type": "writeSelector"
    },
    {
      "Type": "latencySelector"
    }
  ]
}
```

We could omit child selectors, though I would argue that, in the context of logging, more information is better.

Reply:
My concern is actually that the server selector information is a string containing escaped JSON, rather than a JSON object. Consider the example:
```json
{
    "message": "Server selection started",
    "operation": "find",
    "operationId": 1,
    "selector": "{\"Type\":\"compositeSelector\",\"Selectors\":[{\"Type\":\"readPrefSelector\",\"Data\":\"primary\"},{\"Type\":\"latencySelector\"}]}",
    "timestamp": 1692071928112600000,
    "topologyDescription": "Type: Unknown, Servers: [{ Addr: localhost:27017, Type: Unknown, State: Connected, Average RTT: 0s, Min RTT: 0s }, { Addr: localhost:27018, Type: Unknown, State: Connected, Average RTT: 0s, Min RTT: 0s }, { Addr: localhost:27019, Type: Unknown, State: Connected, Average RTT: 0s, Min RTT: 0s }, ]"
}
```
That selector information would be more usable and readable as a JSON object.

Reply:
The [specifications](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#logging) require this field to be a string. This representation seems reasonable to me. However, if you want to unescape the string we can build this data directly, similar to what we do with the [Server](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/description/server.go#L378).

Reply:
Since the spec requires it to be a string, the escaped JSON seems like a reasonable solution.